kolibri/core/content/utils/resource_import.py

```diff
@@ -0,0 +1,513 @@
+
+import requests
+from le_utils.constants import content_kinds
+from six import string_types
+from six import with_metaclass
+
+from kolibri.core.content.errors import InsufficientStorageSpaceError
+from kolibri.core.content.errors import InvalidStorageFilenameError
+from kolibri.core.content.models import ContentNode
+from kolibri.core.content.utils import annotation
+from kolibri.core.content.utils import paths
+from kolibri.core.content.utils.channels import get_mounted_drive_by_id
+from kolibri.core.content.utils.content_manifest import ContentManifest
+from kolibri.core.content.utils.content_manifest import ContentManifestParseError
+from kolibri.core.content.utils.file_availability import LocationError
+from kolibri.core.content.utils.import_export_content import get_import_export_data
+from kolibri.core.content.utils.paths import get_channel_lookup_url
+from kolibri.core.content.utils.paths import get_content_file_name
+from kolibri.core.content.utils.upgrade import get_import_data_for_update
+from kolibri.core.discovery.models import NetworkLocation
+from kolibri.core.discovery.utils.network.client import NetworkClient
+from kolibri.core.discovery.utils.network.errors import NetworkLocationNotFound
+from kolibri.core.tasks.utils import fd_safe_executor
+from kolibri.core.tasks.utils import JobProgressMixin
+from kolibri.utils import conf
+from kolibri.utils import file_transfer as transfer
+from kolibri.utils.system import get_free_space
+
+
+logger = logging.getLogger(__name__)
+
+
+def lookup_channel_listing_status(channel_id, baseurl=None):
+    """
+    Look up the listing status of the channel from the remote, this is surfaced as a
+    `public` boolean field.
+    """
+    resp = requests.get(get_channel_lookup_url(identifier=channel_id, baseurl=baseurl))
+
+    if resp.status_code != 200:
+        return None
+
+    (channel_info,) = resp.json()
+    return channel_info.get("public", None)
+
+
+class ResourceImportManagerBase(with_metaclass(ABCMeta, JobProgressMixin)):
+    public = None
+
+    def __init__(
+        self,
+        channel_id,
+        manifest_file=None,
+        node_ids=None,
+        exclude_node_ids=None,
+        renderable_only=True,
+        fail_on_error=False,
+        content_dir=None,
+    ):
+        self.channel_id = channel_id
+
+        if node_ids is not None:
+            node_ids = set(node_ids)
+
+        if exclude_node_ids is not None:
+            exclude_node_ids = set(exclude_node_ids)
+
+        if manifest_file:
+            if isinstance(manifest_file, string_types):
+                manifest_file = open(manifest_file, "r")
+            content_manifest = ContentManifest()
+            try:
+                content_manifest.read_file(manifest_file)
+                node_ids = content_manifest.get_node_ids_for_channel(channel_id)
+                exclude_node_ids = None
+            except ContentManifestParseError:
+                pass
+            manifest_file.close()
```

Comment:
I think this manifest parsing would be better placed in a factory function that builds a class instance from it, instead of having it in the class constructor. The reason is that it takes precedence on determining the values of `node_ids` and `exclude_node_ids`, but if it were to fail, it makes that codepath a bit ambiguous on the intended outcome since it just suppresses the parse error.

I see an inherited class is also utilizing this argument, so a static class method might be best for the factory. I would expect that the factory function wouldn't accept `node_ids` and `exclude_node_ids` arguments, thereby making the API clear that it's one or the other.